libecole/tests/src/observation/test-milpbipartite.cpp

```diff
@@ -0,0 +1,56 @@
+#include <cstddef>
+#include <iostream>
+
+#include <catch2/catch.hpp>
+#include <xtensor/xmath.hpp>
+#include <xtensor/xview.hpp>
+
+#include "ecole/observation/milpbipartite.hpp"
+
+#include "conftest.hpp"
+#include "observation/unit-tests.hpp"
+
+using namespace ecole;
+
+TEST_CASE("MilpBipartite unit tests", "[unit][obs]") {
+	observation::unit_tests(observation::MilpBipartite{});
+}
+
+TEST_CASE("MilpBipartite return correct observation", "[obs]") {
+	auto normalize = GENERATE(true, false);
+	auto obs_func = observation::MilpBipartite{normalize};
+	auto model = get_model();
+	obs_func.before_reset(model);
+	auto const optional_obs = obs_func.extract(model, false);
+
+	SECTION("Observation is not empty on non terminal state") { REQUIRE(optional_obs.has_value()); }
+
+	SECTION("Observation features are not empty") {
+		auto const& obs = optional_obs.value();
+		REQUIRE(obs.variable_features.size() > 0);
+		REQUIRE(obs.constraint_features.size() > 0);
+		REQUIRE(obs.edge_features.nnz() > 0);
+	}
+
+	SECTION("Observation features have matching shape") {
+		auto const& obs = optional_obs.value();
+		REQUIRE(obs.constraint_features.shape()[0] == obs.edge_features.shape[0]);
+		REQUIRE(obs.variable_features.shape()[0] == obs.edge_features.shape[1]);
+		REQUIRE(obs.edge_features.indices.shape()[0] == 2);
+		REQUIRE(obs.edge_features.indices.shape()[1] == obs.edge_features.nnz());
+	}
+
+	SECTION("Variable features are not all nan") {
```

Comment:
Shouldn't none of them be `nan`?

Reply:
Actually you bring up a good point. Right now I have features `upper_bound` and `lower_bound`, as well as `has_upper_bound` and `has_lower_bound`. If there is no upper bound, I set `has_upper_bound` to zero and `upper_bound` to zero. Should I set the latter to `np.nan` instead?

Or even remove completely the `has_upper_bound` and `has_lower_bound` features since, if I use `np.nan`, it is redundant? The thing is, I expect 99% of users to set `np.nan`'s to zero, in which case users will want to add a new feature mentioning whether the upper bound was `np.nan` in the first place (to let the machine learning model know and not destroy information. So should we already do it for users?

I would be especially curious to know @gasse 's opinion here.

Reply:
Well, inf or -inf would be even cleaner, since those represent correct bounds. Representing those values with an additional flag is required if you want to feed those features to, say, a neural network, otherwise you'll have issues with nans and infs. The question then is whether this should be Ecole's job to do this convertion, or if this should be left to the user ?

Reply:
True, we could have np.inf as well. 

> Representing those values with an additional flag is required if you want to feed those features to, say, a neural network, otherwise you'll have issues with nans and infs. The question then is whether this should be Ecole's job to do this convertion, or if this should be left to the user ?

Yes this is exactly my philosophical question. I was hoping you would have a strong opinion on this matter :-p

Reply:
I don't have strong feelings here, but I tend to prefer the option of using np.inf values, and then letting users doing the pre-processing if they want. This seems less arbitrary, although it will be a little bit less practical for ML users.

Reply:
We should discuss that in our next meeting.

Reply:
The conclusion was to keep it as it is.